extensions-core/s3-extensions/src/main/java/io/druid/storage/s3/S3Utils.java

```diff
@@ -88,121 +84,172 @@ public boolean apply(Throwable e)
   )
   {
-    return new Iterator<StorageObject>()
+    final ListObjectsV2Request request = new ListObjectsV2Request()
+        .withBucketName(bucket)
+        .withPrefix(prefix)
+        .withMaxKeys(numMaxKeys);
+
+    return new Iterator<S3ObjectSummary>()
     {
-      private StorageObjectsChunk objectsChunk;
-      private int objectsChunkOffset;
+      private ListObjectsV2Result result;
+      private Iterator<S3ObjectSummary> objectSummaryIterator;
 
-      @Override
-      public boolean hasNext()
       {
-        if (objectsChunk == null) {
-          objectsChunk = listObjectsChunkedAfter("");
-          objectsChunkOffset = 0;
-        }
-
-        if (objectsChunk.getObjects().length <= objectsChunkOffset) {
-          if (objectsChunk.isListingComplete()) {
-            return false;
-          } else {
-            objectsChunk = listObjectsChunkedAfter(objectsChunk.getPriorLastKey());
-            objectsChunkOffset = 0;
-          }
-        }
+        fetchNextBatch();
+      }
 
-        return true;
+      private void fetchNextBatch()
+      {
+        result = s3Client.listObjectsV2(request);
+        objectSummaryIterator = result.getObjectSummaries().iterator();
+        request.setContinuationToken(result.getContinuationToken());
       }
 
-      private StorageObjectsChunk listObjectsChunkedAfter(final String priorLastKey)
+      @Override
+      public boolean hasNext()
       {
-        try {
-          return retryS3Operation(
-              () -> s3Client.listObjectsChunked(bucket, prefix, null, maxListingLength, priorLastKey)
-          );
-        }
-        catch (Exception e) {
-          throw Throwables.propagate(e);
-        }
+        return objectSummaryIterator.hasNext() || result.isTruncated();
       }
 
       @Override
-      public StorageObject next()
+      public S3ObjectSummary next()
       {
         if (!hasNext()) {
-          throw new IllegalStateException();
+          throw new NoSuchElementException();
         }
-        StorageObject storageObject = objectsChunk.getObjects()[objectsChunkOffset];
-        objectsChunkOffset++;
 
-        return storageObject;
-      }
+        if (objectSummaryIterator.hasNext()) {
+          return objectSummaryIterator.next();
+        }
 
-      @Override
-      public void remove()
-      {
-        throw new UnsupportedOperationException();
-      }
+        if (result.isTruncated()) {
+          fetchNextBatch();
+        }
 
+        if (!objectSummaryIterator.hasNext()) {
+          throw new ISE(
+              "Failed to further iterate on bucket[%s] and prefix[%s]. The last continuationToken was [%s]",
+              bucket,
+              prefix,
+              result.getContinuationToken()
+          );
+        }
 
+        return objectSummaryIterator.next();
+      }
     };
   }
 
-  public static String constructSegmentPath(String baseKey, String storageDir)
+  static String constructSegmentPath(String baseKey, String storageDir)
```

Comment:
we don't know if these were used in extensions, can the things that were public remain so?

Reply:
Reverted.